### ISSUE TITLE
Thread safety?

### ISSUE THREAD

**Issue by hmaarrfk** [cgohlke/tifffile#178]
Is tifffile thread safe in the sense that

1. Open a multi page tiff image with TiffFile in read only mode
2. Access pages from different threads with:
```
tif.pages[index].asarray()
```

Thank you for the advice!

**Comment by cgohlke** [comment]
TiffFile is not thread-safe by default but you can pass a `threading.RLock` instance as `lock` parameter to `TiffPage.asarray()`:

https://github.com/cgohlke/tifffile/blob/08290c7ae44a4248f64d7cd0e74361f570f56e79/tifffile/tifffile.py#L8388-L8438

Setting `tif.filehandle.set_lock(True)` before should also work. The `ZarrStore` should be thread-safe by default.

**Comment by hmaarrfk** [comment]
Thank you very much for your prompt reply. I should have just searched "thread" in the source.

It seems that with specifying an RLock or using tif.filehandle.set_lock, I still get warnings like:

```
tif.pages[index].asarray()
```

The error looks like:
```
tifffile.tifffile:tifffile.py:22909 <tifffile.TiffPage 2 @2621910> <tifffile.TiffTag 771 @2631152> invalid data type 771
```

If i put my own lock around this call, then the error goes away.

I think it is because the pages structure makes many calls to the FileHandle object which are not atomic. 

You've given enough warnings in the code that multi-threaded access is not safe.

In my mind, calling
```
tif.pages[index].asarray()
```
once setting the lock is true should be safe, but maybe there is some larger refactoring to do.

Maybe the TiffFile object should have a similar property like `set_lock`?


I'll let you close or rename the issue as you wish.

**Comment by cgohlke** [comment]
Parsing the TIFF file structure is not thread-safe at all. Only the `TiffPage.asarray` function can be made thread-safe by passing a lock. Only the decoding of large image data without GIL is expected to benefit from multi-threading. Consider disabling internal multi-threading to avoid interference:

Your example becomes:

```python
with lock:
    page = tif.pages[index]
page.asarray(lock=lock, maxworkers=1)
```

**Comment by hmaarrfk** [comment]
Right, so the question is, is it worth doing something like

```
with fh.lock:
    ...
```
in the pages `__getitem__`:
https://github.com/cgohlke/tifffile/blob/master/tifffile/tifffile.py#L7185

**Comment by cgohlke** [comment]
> is it worth doing something like ... in the pages `__getitem__`

I don't see any real benefit.

Just curios: what's the use case for multi-threading `tif.pages[index].asarray()` instead of, for example, `imread(..., key=indices, maxworkers=...)`?

**Comment by hmaarrfk** [comment]
> Just curios

Um, 

1. I didn't know about `imread`'s ability to be multi-threaded.
2. `imread` reopens the file handle every time. Sometimes it is useful to leave the file open (user's can't delete the file afterward).
3. Re the multi-threading. Everybody wants their "library" to be "multi-threaded". Essentially, you get "1 shot" at being multi-threaded, otherwise, the algorithms keep trashing.

I've been creating different wrappers for "numpy-like" arrays that are backed by various files. Then if you give enough information to libraries like `dask`, then they can use their scheduler to make "intelligent" decisions.

**Comment by cgohlke** [comment]
Since you seem to be after efficient reading of whole images from specified single TIFF pages from many threads in parallel, try `imagecodecs.tiff_decode(memory_mapped_file, index=...)`. [Tiff_decode](https://github.com/cgohlke/imagecodecs/blob/master/imagecodecs/_tiff.pyx#L178) is based on the libtiff C library and releases the GIL most of the time throughout the function. I have not tried or benchmarked the memory mapped approach. Is accessing Python's `mmap.mmap` through Cython's typed memoryview thread-safe?

**Comment by hmaarrfk** [comment]
Thanks for the tip.

I've played around with memory mapped files in the past. My conclusion is that the linux kernel (as shipped with ubuntu) doesn't seem to be optimized for that kind of operation.

In terms of optimization, can we set the the alignment of each "page"'s data within the tif file itself? Setting large chunks of data to be aligned to hardware boundaries (I choose 4096) has had the biggest effect in my experiments.
